doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -848,37 +1231,425 @@ The Ltac2 parser can be extended with syntactic notations.
+When parsing, each syntactic class expression returns a value that's bound to a name in the
+notation definition.
+
+Syntactic classes are described with a form of S-expression:
+
+   .. insertprodn ltac2_scope ltac2_scope
+
+   .. prodn::
+      ltac2_scope ::= @string
+      | @int
+      | @name
+      | @name ( {+, @ltac2_scope } )
+
+.. todo no syn class for ints or strings?
+  parm names are not reserved (e.g the var can be named "list1"
+
+Metasyntactic operations that can be applied to other syntactic classes are:
+
+  :n:`opt(@ltac2_scope)`
+    Parses an optional :token:`ltac2_scope`.  The associated value is either :n:`None` or
+    enclosed in :n:`Some`
+
+  :n:`list1(@ltac2_scope {? , @string })`
+    Parses a list of one or more :token:`ltac2_scope`\s.  If :token:`string` is specified,
+    items must be separated by :token:`string`.
+
+  :n:`list0(@ltac2_scope {? , @string })`
+    Parses a list of zero or more :token:`ltac2_scope`\s.  If :token:`string` is specified,
+    items must be separated by :token:`string`.  For zero items, the associated value
+    is an empty list.
+
+  :n:`seq({+, @ltac2_scope })`
+    Parses the :token:`ltac2_scope`\s in order.  The associated value is a tuple,
+    omitting :token:`ltac2_scope`\s that are :token:`string`\s.
+    `self` and `next` are not permitted within `seq`.
+
+The following classes represent nonterminals with some special handling.  The
+table further down lists the classes that that are handled plainly.
+
+  :n:`constr {? ( @scope_key ) }`
+    Parses a :token:`term`.  If specified, the :token:`scope_key` is used to interpret
+    the term (as described in  :ref:`LocalInterpretationRulesForNotations`)
+
+    .. todo PR the original text describes "constr:(c%A%B)" which I couldn't make sense of
+       Is that something we should keep (and if so, what is an example?)
+
+  :n:`open_constr`
+    Parses an open :token:`term`.
+
+  :n:`ident`
+    Parses :token:`ident` or :n:`$@ident`.  The first form returns :n:`ident:(@ident)`,
+    while the latter form returns the variable :n:`@ident`.
+
+  :n:`tactic {? (@int) }`
+    Parses an :token:`ltac2_expr`.  If :token:`int` is specified, the construct
+    parses a :n:`ltac2_expr@int`, for example `tactic(5)` parses :token:`ltac2_expr5`.
+    `tactic(6)` parses :token:`ltac2_expr`.
+    :token:`int` must be in the range `0 .. 6`.
+
+    .. todo this doesn't work as expected: "::" is in ltac2_expr1
+       Ltac2 Notation "ex4" x(tactic(0)) := x.
+       ex4 auto :: [auto].
+
+  .. not sure "self" and "next" do anything special.  I get the same error
+     message for both from constructs like
+
+     Ltac2 Notation "ex5" x(self) := auto.
+     ex5 match.
+
+     Syntax error: [tactic:tac2expr level 5] expected after 'match' (in [tactic:tac2expr]).
+
+  :n:`self`
+    parses an Ltac2 expression at the current level and returns it as is.
+
+  :n:`next`
+    parses an Ltac2 expression at the next level and returns it as is.
```

Comment:
Next level up, or next level down?  (And which is tighter?)

Reply:
(I think it's next level down, and tighter, but I'm not sure, and in any case this should be documented, either here or in the Gallina notations section (if it's not already) or both)

Reply:
I think these are equivalent to the parser's SELF and NEXT but now looking at it again I don't see how that makes any sense.  @ppedrot can you give an example showing how to use it?

Reply:
It's the direct equivalent of camlp5 NEXT / SELF, so it makes sense w.r.t. the current tactic level being expanded. Not sure whether I have a practical example at hand though.

Reply:
Hmmm, this feature seems to be sort-of broken anyway, see https://github.com/coq/coq/issues/11866 and https://github.com/coq/coq/issues/12807 and https://github.com/coq/coq/issues/12808

Reply:
SELF and NEXT only make sense in the context of defining productions for some nonterminal.  Notations with precedence number would already substitute SELF/NEXT for the variables in the expression.  Seems like this may have been added without thinking it through, especially if you can't provide an example.